libs/ngx-nuts-and-bolts/src/lib/table-state/table-state.helper.ts

```diff
@@ -0,0 +1,153 @@
+				pageSize,
+			};
+		}),
+		distinctUntilChanged(
+			(pageEvent1, pageEvent2) =>
+				pageEvent1.pageIndex === pageEvent2.pageIndex && pageEvent1.pageSize === pageEvent2.pageSize
+		)
+	);
+}
+
+export function createSortingObservable(route: ActivatedRoute): Observable<ISortInfo> {
+	return route.queryParamMap.pipe(
+		map((paramMap) => {
+			const sortingKey = paramMap.get(TableQueryParam.SORT_KEY) ?? '';
+			const sortDirection = (paramMap.get(TableQueryParam.SORT_DIRECTION) ?? '') as Sort;
+
+			return {
+				sortingKey,
+				sortDirection,
+			};
+		}),
+		distinctUntilChanged(isEqual)
+	);
+}
+
+export function createCustomFiltersObservable<TFilterValue extends Record<string, unknown> | null = null>(
+	route: ActivatedRoute,
+	router: Router
+): Observable<TFilterValue | null> {
+	return route.queryParamMap.pipe(
+		map((queryParamMap) => {
+			let filters = null;
+
+			try {
+				if (queryParamMap.has(TableQueryParam.CUSTOM_FILTERS)) {
+					filters = JSON.parse(atob(queryParamMap.get(TableQueryParam.CUSTOM_FILTERS) || ''));
+				}
+			} catch (e) {
+				console.error(e);
+			}
+
+			if (filters && isFilterEmpty(filters)) {
+				clearFilters(router);
+				return null;
+			}
+
+			return filters;
+		}),
+		distinctUntilChanged(isEqual)
+	);
+}
+
+function isFilterEmpty<TFilterValue extends Record<string, unknown> | null = null>(filterValue: TFilterValue): boolean {
+	if (!filterValue) {
+		return true;
+	}
+
+	const filterHasSomeValue = Object.values(filterValue).some((value) => {
+		if (typeof value === 'string') {
+			return Boolean(value);
+		} else if (typeof value === 'number') {
+			return value || value === 0;
+		} else if (value instanceof Array) {
+			return value.length;
+		}
+
+		return Boolean(value);
+	});
+
+	return !filterHasSomeValue;
+}
+
+export function onPageChangeHelper(router: Router, pageInfo: IPageInfo): void {
+	router.navigate([], {
+		replaceUrl: true,
+		queryParamsHandling: 'merge',
+		queryParams: {
+			[TableQueryParam.PAGE_INDEX]: pageInfo.pageIndex || null,
+			[TableQueryParam.PAGE_SIZE]: pageInfo.pageSize,
+		},
+	});
+}
+
+export function onSortChangeHelper(router: Router, sort: ISortInfo): void {
+	const sortDirection = sort.sortDirection || null;
+	const sortKey = sortDirection ? sort.sortingKey : null;
+	router.navigate([], {
+		replaceUrl: true,
+		queryParamsHandling: 'merge',
+		queryParams: {
+			[TableQueryParam.SORT_KEY]: sortKey,
+			[TableQueryParam.SORT_DIRECTION]: sortDirection,
+		},
+	});
+}
+
+export function onFiltersChange<TFilterValue extends Record<string, unknown> | null = null>(
+	router: Router,
+	customFilters: TFilterValue
+): void {
+	router.navigate([], {
+		replaceUrl: true,
+		queryParamsHandling: 'merge',
+		queryParams: {
+			[TableQueryParam.CUSTOM_FILTERS]: customFilters ? btoa(JSON.stringify(customFilters)) : null,
+			[TableQueryParam.PAGE_INDEX]: null, // go to first page on filters change
+		},
+	});
+}
+
+export function clearFilters(router: Router): void {
```

Comment:
Due to nature of promises, there is literally no reason NOT to return it here, unlike observables there is no need for a subscription down the line, yet it would allow consumers to actually handle navigation result as they like.

Reply:
@JosipJanus please check all the places where you call methods on router and return whatever they return.